test/exchange/EndemicPrivateSale.js

```
@@ -0,0 +1,180 @@
+  }
+
+  async function deploy() {
+    [owner] = await ethers.getSigners();
+
+    const result = await deployEndemicExchangeWithDeps();
+
+    endemicExchange = result.endemicExchangeContract;
+
+    nftContract = (await deployEndemicCollectionWithFactory()).nftContract;
+
+    await mintERC721(owner.address);
+  }
+
+  describe('Initial State', function () {
+    it('should set domain separator', async function () {
+      await deploy();
+
+      expect(await endemicExchange.DOMAIN_SEPARATOR()).to.exist;
+    });
+  });
+
+  describe('Buy from private sale', function () {
+    beforeEach(deploy);
+
+    it('should fail with private sale expired', async function () {
+      await expect(
+        endemicExchange.buyFromPrivateSale(
+          nftContract.address,
+          1,
+          ZERO_ONE_ETHER,
+          LAST_YEAR_TIMESTAMP,
+          RANDOM_V_VALUE,
+          RANDOM_R_VALUE,
+          RANDOM_S_VALUE
+        )
+      ).to.be.revertedWith(PRIVATE_SALE_EXPIRED);
+    });
+
+    it('should fail with price not correct (too much provided)', async function () {
+      await expect(
+        endemicExchange.buyFromPrivateSale(
+          nftContract.address,
+          1,
+          ZERO_ONE_ETHER,
+          RANDOM_TIMESTAMP,
+          RANDOM_V_VALUE,
+          RANDOM_R_VALUE,
+          RANDOM_S_VALUE,
+          {
+            value: ONE_ETHER,
+          }
+        )
+      ).to.be.revertedWith(PRICE_NOT_CORRECT);
+    });
+
+    it('should fail with price not correct (too low provided)', async function () {
+      await expect(
+        endemicExchange.buyFromPrivateSale(
+          nftContract.address,
+          1,
+          ONE_ETHER,
+          RANDOM_TIMESTAMP,
+          RANDOM_V_VALUE,
+          RANDOM_R_VALUE,
+          RANDOM_S_VALUE,
+          {
+            value: ZERO_ONE_ETHER,
+          }
+        )
+      ).to.be.revertedWith(PRICE_NOT_CORRECT);
+    });
+
+    it('should fail with invalid signature', async function () {
```

Let's add tests:
- should fail to buy with valid signature and invalid buyer (I can't buy private sale that was meant for you)